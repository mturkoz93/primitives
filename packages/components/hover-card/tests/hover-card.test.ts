import { afterEach, beforeEach, describe, expect, it, test, vitest } from 'vitest'
import type { VueWrapper } from '@vue/test-utils'
import { enableAutoUnmount, mount } from '@vue/test-utils'
import { axe } from 'vitest-axe'
import { nextTick } from 'vue'
import {
  OkuHoverCard,
  OkuHoverCardArrow,
  OkuHoverCardContent,
} from '../src'

import AsyncUpdateVue from '../src/stories/AsyncUpdate.vue'
import AnimatedVue from '../src/stories/Animated.vue'
import ContainTextSelectionVue from '../src/stories/ContainTextSelection.vue'
import ControlledVue from '../src/stories/Controlled.vue'
import ForcedMountVue from '../src/stories/ForcedMount.vue'
import NonPortalVue from '../src/stories/NonPortal.vue'

enableAutoUnmount(afterEach)

describe('OkuHoverCard', () => {
  it('OkuHoverCard renders correctly', () => {
    const wrapper = mount(OkuHoverCard)
    expect(wrapper.html()).toBe('')
  })

  it('OkuHoverCardArrow renders correctly', () => {
    const spy = vitest.spyOn(global.console, 'warn').mockImplementation(() => { })
    const wrapper = () => mount(OkuHoverCardArrow)
    expect(() => wrapper()).toThrowError('`OkuPopperArrow` must be used within `OkuPopperContent`') // TODO
    expect(spy).toHaveBeenCalled()

    expect(spy.mock.calls[0][0]).toContain('[Vue warn]: injection "Symbol(OkuPopperContent)" not found.')
  })

  it('OkuHoverCardContent renders correctly', () => {
    const spy = vitest.spyOn(global.console, 'warn').mockImplementation(() => { })
    const wrapper = () => mount(OkuHoverCardContent)
    expect(() => wrapper()).toThrowError('`OkuHoverCardContent` must be used within `OkuHoverCard`') // TODO
    expect(spy).toHaveBeenCalled()

    expect(spy.mock.calls[0][0]).toContain('[Vue warn]: injection "Symbol(OkuHoverCardPortal)" not found.')
  })
})

describe('Animated', () => {
  let wrapper: VueWrapper<InstanceType<typeof AnimatedVue>>

  beforeEach(async () => {
    wrapper = mount(AnimatedVue, {
      attachTo: document.body,
    })
    const button = wrapper.find('[class="hover-card-trigger"]')
    button.trigger('click')
  })

  it('renders correctly', () => {
    expect(document.body.innerHTML).toMatchSnapshot()
  })

  it('should be able to close', async () => {
    const _element = wrapper.find('[class="hover-card-trigger"]')
    _element.trigger('pointerleave')
    await nextTick()
    expect(document.body.innerHTML).toMatchSnapshot()
  })
})

describe('AsyncUpdate', async () => {
  let wrapper: VueWrapper<InstanceType<typeof AsyncUpdateVue>>

  beforeEach(async () => {
    wrapper = mount(AsyncUpdateVue, {
      attachTo: document.body,
    })

    await wrapper.find('[class="hover-card-trigger"]').trigger('pointerenter')
    await new Promise(resolve => setTimeout(resolve, 500))
    await nextTick()
  })

  /**
    * @vitest-environment jsdom
    */
  it('axe accessibility tests', async () => {
    expect(await axe(wrapper.element)).toHaveNoViolations()
  })

  test('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be able to close', async () => {
    await wrapper.find('[class="hover-card-trigger"]').trigger('pointerleave')
    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('ContainTextSelection', async () => {
  let wrapper: VueWrapper<InstanceType<typeof ContainTextSelectionVue>>

  beforeEach(async () => {
    wrapper = mount(ContainTextSelectionVue, {
      attachTo: document.body,
    })

    await wrapper.find('[class="hover-card-trigger"]').trigger('pointerenter')
    await nextTick()
  })

  /**
        * @vitest-environment jsdom
        */
  it('axe accessibility tests', async () => {
    expect(await axe(wrapper.element)).toHaveNoViolations()
  })

  test('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be able to close', async () => {
    await wrapper.find('[class="hover-card-trigger"]').trigger('pointerleave')
    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Controlled', async () => {
  let wrapper: VueWrapper<InstanceType<typeof ControlledVue>>

  beforeEach(async () => {
    wrapper = mount(ControlledVue, {
      attachTo: document.body,
    })

    await wrapper.find('[class="hover-card-trigger"]').trigger('pointerenter')
    await nextTick()
  })

  /**
     * @vitest-environment jsdom
     */
  it('axe accessibility tests', async () => {
    expect(await axe(wrapper.element)).toHaveNoViolations()
  })

  test('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be able to close', async () => {
    await wrapper.find('[class="hover-card-trigger"]').trigger('pointerleave')
    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('ForcedMount', async () => {
  let wrapper: VueWrapper<InstanceType<typeof ForcedMountVue>>

  beforeEach(async () => {
    wrapper = mount(ForcedMountVue, {
      attachTo: document.body,
    })
    await nextTick()
  })

  /**
    * @vitest-environment jsdom
    */
  it('axe accessibility tests', async () => {
    expect(await axe(wrapper.element)).toHaveNoViolations()
  })

  test('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be able to close', async () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('NonPortal', async () => {
  let wrapper: VueWrapper<InstanceType<typeof NonPortalVue>>

  beforeEach(async () => {
    wrapper = mount(NonPortalVue, {
      attachTo: document.body,
    })
    await nextTick()
  })

  /**
      * @vitest-environment jsdom
      */
  it('axe accessibility tests', async () => {
    expect(await axe(wrapper.element)).toHaveNoViolations()
  })

  test('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be able to close', async () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
