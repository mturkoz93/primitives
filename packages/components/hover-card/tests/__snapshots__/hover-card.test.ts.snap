// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Animated > renders correctly 1`] = `"<div data-v-app=\\"\\"><div style=\\"padding: 50px; display: flex; justify-content: center;\\"><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> trigger </a><!----></div></div>"`;

exports[`Animated > should be able to close 1`] = `"<div data-v-app=\\"\\"></div><div data-v-app=\\"\\"><div style=\\"padding: 50px; display: flex; justify-content: center;\\"><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> trigger </a><!----></div></div>"`;

exports[`AsyncUpdate > renders correctly 1`] = `
"<div style=\\"padding: 50px; display: flex; justify-content: center;\\"><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> trigger </a>
  <!---->
</div>"
`;

exports[`AsyncUpdate > should be able to close 1`] = `
"<div style=\\"padding: 50px; display: flex; justify-content: center;\\"><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> trigger </a>
  <!---->
</div>"
`;

exports[`ContainTextSelection > renders correctly 1`] = `
"<div style=\\"padding: 50px; display: flex; justify-content: center; flex-direction: column; align-items: center;\\">
  <div style=\\"display: flex; gap: 30px;\\"><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> single </a>
    <!----><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> nested </a>
    <!---->
  </div>
  <div style=\\"max-width: 800px;\\">
    <p> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer feugiat mattis malesuada. Fusce elementum vulputate aliquet. Integer fringilla porta eros. Ut ultricies mattis nisi. Sed et tempor massa. Sed non arcu ut velit scelerisque bibendum tempor sed mi. In non consequat sapien. Donec sollicitudin eget tellus ut venenatis. Donec posuere sem ante, nec iaculis arcu varius sit amet. Praesent non tortor quam. Curabitur dapibus justo a commodo ornare. </p>
    <p> Suspendisse eleifend consequat iaculis. Nunc bibendum velit felis, nec vulputate purus egestas quis. Integer mauris dui, pulvinar non metus id, tristique dignissim elit. Vivamus massa tellus, porttitor id lorem non, molestie aliquam dolor. Pellentesque erat quam, pellentesque non metus id, tempus sagittis massa. </p>
    <p> Sed at elementum sem, non venenatis leo. Ut vulputate consectetur finibus. Sed nunc lectus, accumsan in nisl et, vehicula pretium nisi. Vivamus vestibulum ante quis urna consequat, ultrices condimentum sem commodo. Pellentesque eget orci laoreet, feugiat purus sed, maximus nisi. Suspendisse commodo venenatis facilisis. </p>
  </div>
</div>"
`;

exports[`ContainTextSelection > should be able to close 1`] = `
"<div style=\\"padding: 50px; display: flex; justify-content: center; flex-direction: column; align-items: center;\\">
  <div style=\\"display: flex; gap: 30px;\\"><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> single </a>
    <!----><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> nested </a>
    <!---->
  </div>
  <div style=\\"max-width: 800px;\\">
    <p> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer feugiat mattis malesuada. Fusce elementum vulputate aliquet. Integer fringilla porta eros. Ut ultricies mattis nisi. Sed et tempor massa. Sed non arcu ut velit scelerisque bibendum tempor sed mi. In non consequat sapien. Donec sollicitudin eget tellus ut venenatis. Donec posuere sem ante, nec iaculis arcu varius sit amet. Praesent non tortor quam. Curabitur dapibus justo a commodo ornare. </p>
    <p> Suspendisse eleifend consequat iaculis. Nunc bibendum velit felis, nec vulputate purus egestas quis. Integer mauris dui, pulvinar non metus id, tristique dignissim elit. Vivamus massa tellus, porttitor id lorem non, molestie aliquam dolor. Pellentesque erat quam, pellentesque non metus id, tempus sagittis massa. </p>
    <p> Sed at elementum sem, non venenatis leo. Ut vulputate consectetur finibus. Sed nunc lectus, accumsan in nisl et, vehicula pretium nisi. Vivamus vestibulum ante quis urna consequat, ultrices condimentum sem commodo. Pellentesque eget orci laoreet, feugiat purus sed, maximus nisi. Suspendisse commodo venenatis facilisis. </p>
  </div>
</div>"
`;

exports[`Controlled > renders correctly 1`] = `
"<button> Toggle </button>
<div style=\\"padding: 50px; display: flex; justify-content: center;\\"><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> trigger </a>
  <!---->
</div>"
`;

exports[`Controlled > should be able to close 1`] = `
"<button> Toggle </button>
<div style=\\"padding: 50px; display: flex; justify-content: center;\\"><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> trigger </a>
  <!---->
</div>"
`;

exports[`ForcedMount > renders correctly 1`] = `
"<div style=\\"padding: 50px; display: flex; justify-content: center;\\"><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> triggers </a>
  <!--teleport start-->
  <!--teleport end-->
</div>"
`;

exports[`ForcedMount > should be able to close 1`] = `
"<div style=\\"padding: 50px; display: flex; justify-content: center;\\"><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> triggers </a>
  <!--teleport start-->
  <!--teleport end-->
</div>"
`;

exports[`NonPortal > renders correctly 1`] = `
"<div><button>button</button><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> trigger </a>
  <!----><button>button</button>
</div>"
`;

exports[`NonPortal > should be able to close 1`] = `
"<div><button>button</button><a data-state=\\"closed\\" href=\\"/\\" class=\\"hover-card-trigger\\"> trigger </a>
  <!----><button>button</button>
</div>"
`;
